src/setup/Agda/Setup.hs

```diff
@@ -0,0 +1,147 @@
+
+import           Data.ByteString            ( ByteString )
+import qualified Data.ByteString            as BS
+import           Data.Functor               ( (<&>) )
+import           Data.List                  ( intercalate )
+
+import           Language.Haskell.TH.Syntax ( qAddDependentFile, runIO )
+
+-- Import instance Lift ByteString if not supplied by bytestring.
+#if !MIN_VERSION_bytestring(0,11,2)
+import           Instances.TH.Lift          ()
+#endif
+
+import           System.Directory
+  ( XdgDirectory (..)
+  , canonicalizePath, createDirectoryIfMissing, doesDirectoryExist
+  , getAppUserDataDirectory, getXdgDirectory, removeFile
+  )
+import           System.Environment         ( lookupEnv )
+import           System.FileLock            ( pattern Exclusive, withFileLock )
+import           System.FilePath            ( (</>), joinPath, splitFileName, takeFileName )
+
+import           Agda.Setup.DataFiles       ( dataFiles, dataPath )
+import           Agda.VersionCommit         ( versionWithCommitInfo )
+
+-- | What to append to the @AGDA_DIR@ to construct the Agda data directory.
+
+dataDirNameParts :: [FilePath]
+dataDirNameParts = [ "share", versionWithCommitInfo ]
+
+-- | Given the `AGDA_DIR`, what should the Agda data dir be?
+
+mkDataDir :: FilePath -> FilePath
+mkDataDir base = joinPath $ base : dataDirNameParts
+
+
+-- Tell TH that all the dataFiles are needed for compilation.
+[] <$ mapM_ (qAddDependentFile . dataPath) dataFiles
+
+-- | The embedded contents of the Agda data directory,
+--   generated by Template Haskell at compile time.
+
+embeddedDataDir :: [(FilePath, ByteString)]
+embeddedDataDir = $(do
+
+    -- Load all the dataFiles.
+    contents <- runIO $ mapM (BS.readFile . dataPath) dataFiles
+
+    -- Return the association list as Exp.
+    [| zip dataFiles contents |]
+  )
+
+-- | Get the path to @~/.agda@ (system-specific).
+--   Can be overwritten by the @AGDA_DIR@ environment variable.
+--
+--   (This is not to be confused with the directory 'getDataDir' for the data files
+--   that Agda needs (e.g. the primitive modules).)
+--
+getAgdaAppDir :: IO FilePath
+getAgdaAppDir = do
+  -- The default can be overwritten by setting the AGDA_DIR environment variable
+  lookupEnv "AGDA_DIR" >>= \case
+    Nothing -> agdaDir
+    Just dir -> doesDirectoryExist dir >>= \case
+      True  -> canonicalizePath dir
+      False -> do
+        d <- agdaDir
+        putStrLn $ "Warning: Environment variable AGDA_DIR points to non-existing directory " ++ show dir ++ ", using " ++ show d ++ " instead."
+        return d
+  where
+    -- System-specific command to build the path to ~/.agda (Unix) or %APPDATA%\agda (Win)
+    agdaDir = do
+      legacyAgdaDir <- getAppUserDataDirectory "agda"
+      doesDirectoryExist legacyAgdaDir >>= \case
+        True  -> return legacyAgdaDir
+        False -> getXdgDirectory XdgConfig "agda"
+
+-- | This overrides the 'getDataDir' from ''Paths_Agda''.
```

Comment:
I'm not sure putting the data directory in the app directory makes a ton of sense: Nix/Guix users will now have two copies of the primitive modules and interfaces: one in the read-only store (needed to build e.g. the standard library) and one in their home directory (generated by Agda at runtime). This also means that every invocation of agda in a Nix build must override `AGDA_DIR` to some location in the store, which feels weird.

Was there a good reason for this change?

Reply:
From https://github.com/agda/agda/issues/7712:

> The main reason is that it's difficult to correctly distribute the data files with Agda, due to the odd dependence on the Cabal variable `Agdadata_dir` and various other packaging woes.

I am not sure what this means. Can you elaborate? What is odd about `Agda_datadir`?

> and paves the way for dropping v1-cabal.

Could you explain how?

Reply:
> What is odd about Agda_datadir?

If you download the Agda executable, e.g. from the CI artifacts, you can't run it without setting the `Agda_datadir`, because it points to a specific location _where Agda was built_.

> paves the way for dropping v1-cabal.

This is more of a developer concern.  Basically, the executable created by `cabal v2-build` is now usable out of the box.

Reply:
I suppose that makes life simpler for some people.

Can we at least decouple the data directory from the AGDA_DIR? I really don't think it belongs in there. My suggestion would be to default to `Agda_datadir` is that is set, and otherwise use something like `$XDG_DATA_HOME/agda/version[-hash]`. This would solve the issue with Nix/Guix builds (we can set `Agda_datadir` in the wrapper) and also not pollute `$XDG_CONFIG_HOME` with data files. I can work on a PR if that sounds good.